src/main/java/org/folio/bulkops/service/PreviewService.java

```diff
@@ -124,6 +139,87 @@ public UnifiedTable getPreview(BulkOperation operation, BulkOperationStep step,
     };
   }
 
+  private UnifiedTable buildCompositePreview(BulkOperation bulkOperation, int offset, int limit) {
+    var csvTable = buildPreviewFromCsvFile(bulkOperation.getLinkToMatchedRecordsCsvFile(), Instance.class, offset, limit, bulkOperation);
+    if (isNotEmpty(bulkOperation.getLinkToModifiedRecordsMarcFile())) {
+      referenceProvider.updateMappingRules();
+      var rules = ruleService.getMarcRules(bulkOperation.getId());
+      var changedOptionsSet = getChangedOptionsSet(rules);
+      var compositeTable = UnifiedTableHeaderBuilder.getEmptyTableWithHeaders(Instance.class);
+      noteTableUpdater.extendTableWithInstanceNotesTypes(compositeTable, changedOptionsSet);
+      var sourcePosition = getCellPositionByName(INSTANCE_SOURCE);
+      var hridPosition = getCellPositionByName(INSTANCE_HRID);
+      var administrativeNotesPosition = getCellPositionByName(INSTANCE_ADMINISTRATIVE_NOTE);
+      var staffSuppressPosition = getCellPositionByName(INSTANCE_STAFF_SUPPRESS);
+      var headers = compositeTable.getHeader().stream()
+        .map(Cell::getValue)
+        .toList();
+      csvTable.getRows().forEach(csvRow -> {
+        if (FOLIO.equals(csvRow.getRow().get(sourcePosition))) {
+          compositeTable.addRowsItem(csvRow);
+        } else {
+          findMarcRecordByHrid(bulkOperation, csvRow.getRow().get(hridPosition)).ifPresent(marcRecord -> {
+            var marcRow = new Row().row(marcToUnifiedTableRowMapper.processRecord(marcRecord, headers));
+            marcRow.getRow().set(administrativeNotesPosition, csvRow.getRow().get(administrativeNotesPosition));
+            marcRow.getRow().set(staffSuppressPosition, csvRow.getRow().get(staffSuppressPosition));
+            compositeTable.addRowsItem(marcRow);
+          });
+        }
+      });
+      return compositeTable;
+    }
+    return csvTable;
+  }
+
+  private void enrichMarcWithAdministrativeData(UnifiedTable unifiedTable, BulkOperation bulkOperation) {
+    var hridPosition = getCellPositionByName(INSTANCE_HRID);
+    var administrativeNotesPosition = getCellPositionByName(INSTANCE_ADMINISTRATIVE_NOTE);
+    var staffSuppressPosition = getCellPositionByName(INSTANCE_STAFF_SUPPRESS);
+      unifiedTable.getRows().forEach(row -> findInstanceByHrid(bulkOperation, row.getRow().get(hridPosition)).ifPresent(instance -> {
+        row.getRow().set(staffSuppressPosition, isNull(instance.getStaffSuppress()) ? EMPTY : instance.getStaffSuppress().toString());
+        row.getRow().set(administrativeNotesPosition, String.join(ITEM_DELIMITER_SPACED, instance.getAdministrativeNotes()));
+      }));
+  }
+
+  private Optional<Instance> findInstanceByHrid(BulkOperation bulkOperation, String hrid) {
```

Comment:
Is it possible to find all instances by set of hrids in one operation ?

Reply:
good catch, thanks!